src/winshellext/F3DThumbnailProvider.cxx

```diff
@@ -151,42 +153,110 @@ IFACEMETHODIMP F3DThumbnailProvider::Initialize(LPCWSTR pszFilePath, DWORD)
 // Generate the thumbnail bitmap for the requested file.
 IFACEMETHODIMP F3DThumbnailProvider::GetThumbnail(UINT cx, HBITMAP* phbmp, WTS_ALPHATYPE* pdwAlpha)
 {
-  HRESULT ret = E_FAIL;
-  wchar_t warg[MAX_PATH * 2];
-  swprintf_s(warg, MAX_PATH * 2,
-    L"f3d --dry-run --quiet -sta --no-background --resolution=%d,%d \"%s\"",
-    cx, cx, m_filePath);
-  int argc;
-  char** argv;
-  ::ConvertWCommandLineToArgs(warg, argc, argv);
-
-  try
+  // Get a temporary PNG image file name
+  wchar_t lpTempPathBuffer[MAX_PATH];
+  wchar_t image_filename[MAX_PATH];
+
+  DWORD dwRetVal = GetTempPathW(MAX_PATH, lpTempPathBuffer);
+  if (dwRetVal > MAX_PATH || (dwRetVal == 0))
   {
-#if NDEBUG
-    vtkObject::GlobalWarningDisplayOff();
-#endif
+    return HRESULT_FROM_WIN32(GetLastError());
+  }
 
-    F3DLoader loader;
-    vtkNew<vtkImageData> image;
-    int r = loader.Start(argc, argv, image);
-    UINT imgWidth = static_cast<UINT>(image->GetDimensions()[0]);
+  if (GetTempFileName(lpTempPathBuffer, L"f3d", 0, image_filename) == 0)
+  {
+    return HRESULT_FROM_WIN32(GetLastError());
+  }
 
-    if (imgWidth == cx && r == EXIT_SUCCESS)
-    {
-      ::CreateBitmapFromBuffer(image, phbmp);
-      *pdwAlpha = WTSAT_ARGB;
-      ret = NOERROR;
-    }
+  wcscat_s(image_filename, L".png");
+
+  // Create command to run
+  wchar_t command[MAX_PATH * 3 + 20];
+  swprintf_s(command, MAX_PATH * 3 + 20,
+    L"\"%s\" --input \"%s\" --output \"%s\" --dry-run -sta --no-background --quiet --resolution %d,%d\"",
+    m_f3dPath, m_filePath, image_filename, cx, cx);
+
+  // Let's launch the process
+  PROCESS_INFORMATION pi;
+  ZeroMemory(&pi, sizeof(PROCESS_INFORMATION));
+
+  STARTUPINFO si;
+  ZeroMemory(&si, sizeof(STARTUPINFO));
+  si.cb = sizeof(STARTUPINFO);
+  si.dwFlags |= STARTF_USESTDHANDLES;
+
+  if (!CreateProcess(
+        m_f3dPath, command, nullptr, nullptr, TRUE, CREATE_NO_WINDOW, nullptr, nullptr, &si, &pi))
+  {
+    return HRESULT_FROM_WIN32(GetLastError());
+  }
+
+  WaitForSingleObject(pi.hProcess, INFINITE);
```

Comment:
Maybe we can add a reasonable timeout? If the user has a very big STEP file for instance, it will take forever to render, so maybe it's better to skip it?

Reply:
So we have to decide the timeout value :) 5 secs?

Reply:
5 secs seems a bit too much 2/3 secs seems more reasonnable.

Reply:
I would say the opposite, 10 secs :)

Reply:
haha, @jpouderoux do you have an idea on how long WaterBottle takes ?

Reply:
I propose to set a 8sec timeout by default. It can be adjusted by a new CMake option.

Reply:
Sound good to me

Reply:
Fair enough.